hummingbot/connector/derivative/hook_odyssey_perpetual/hook_odyssey_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,122 @@
+from hummingbot.core.data_type.perpetual_api_order_book_data_source import PerpetualAPIOrderBookDataSource
+from hummingbot.core.event.event_forwarder import EventForwarder
+from hummingbot.core.event.events import OrderBookEvent
+from hummingbot.core.web_assistant.ws_assistant import WSAssistant
+from hummingbot.logger import HummingbotLogger
+
+if TYPE_CHECKING:
+    from hummingbot.connector.derivative.hook_odyssey_perpetual.hook_odyssey_perpetual_derivative import (
+        HookOdysseyPerpetualDerivative,
+    )
+
+
+class HookOdysseyPerpetualAPIOrderBookDataSource(PerpetualAPIOrderBookDataSource):
+    _hpopobds_logger: Optional[HummingbotLogger] = None
+
+    def __init__(
+        self,
+        trading_pairs: List[str],
+        connector: "HookOdysseyPerpetualDerivative",
+        data_source: HookOdysseyPerpetualDataSource,
+        domain: str,
+    ):
+        super().__init__(trading_pairs)
+        self._connector = connector
+        self._domain = domain
+        self._data_source = data_source
+        self._forwarders = []
+        self._configure_event_forwarders()
+
+    async def get_last_traded_prices(
+        self, trading_pairs: List[str], domain: Optional[str] = None
+    ) -> Dict[str, float]:
+        return await self._connector.get_last_traded_prices(trading_pairs=trading_pairs)
+
+    async def listen_for_subscriptions(self):
+        # The socket reconnection is managed by the data_source. This method should do nothing
+        pass
+
+    async def _parse_order_book_snapshot_message(
+        self, raw_message: OrderBookMessage, message_queue: asyncio.Queue
+    ):
+        # In HookOdysseyPerpetual, 'raw_message' is the OrderBookMessage from the data source
+        message_queue.put_nowait(raw_message)
+
+    async def _connected_websocket_assistant(self) -> WSAssistant:
+        # HookOdysseyPerpetual uses GraphQL websockets to consume stream events
+        raise NotImplementedError
+
+    async def _subscribe_channels(self, ws: WSAssistant):
+        # HookOdysseyPerpetual uses GraphQL websockets to consume stream events
+        raise NotImplementedError
+
+    async def _order_book_snapshot(self, trading_pair: str) -> OrderBookMessage:
+        return await self._data_source.get_order_book_snapshot(trading_pair)
+
+    async def _parse_trade_message(
+        self, raw_message: OrderBookMessage, message_queue: asyncio.Queue
+    ):
+        # In HookOdysseyPerpetual, 'raw_message' is the OrderBookMessage from the data source
+        message_queue.put_nowait(raw_message)
+
+    async def _parse_order_book_diff_message(
+        self, raw_message: OrderBookMessage, message_queue: asyncio.Queue
+    ):
+        # In HookOdysseyPerpetual, 'raw_message' is the OrderBookMessage from the data source
+        message_queue.put_nowait(raw_message)
+
+    async def get_funding_info(self, trading_pair: str) -> FundingInfo:
+        index_price = await self._data_source.get_last_traded_price(trading_pair)
```

Comment:
do we need both here?

Reply:
The FundingInfo object has a field for both mark price and index price. This line is just getting the cached index price to include along with mark, etc.